libminifi/include/utils/FlatMap.h

```diff
@@ -0,0 +1,270 @@
+
+   private:
+    typename Container::const_iterator it_;
+  };
+
+  FlatMap() = default;
+  FlatMap(const FlatMap&) = default;
+  FlatMap(FlatMap&&) noexcept = default;
+  FlatMap(std::initializer_list<value_type> items) : data_(items) {}
+  template<class InputIterator>
+  FlatMap(InputIterator begin, InputIterator end) : data_(begin, end) {}
+
+  FlatMap& operator=(const FlatMap& source) = default;
+  FlatMap& operator=(FlatMap&& source) = default;
+  FlatMap& operator=(std::initializer_list<value_type> items) {
+    data_ = items;
+    return *this;
+  }
+
+  size_type size() const noexcept {
+    return data_.size();
+  }
+
+  V& operator[](const K& key) {
+    auto it = find(key);
+    if (it != end()) {
+      return it->second;
+    }
+    data_.emplace_back(key, V{});
+    return data_.rbegin()->second;
+  }
+
+  iterator erase(const_iterator pos) {
+    auto offset = pos.it_ - data_.begin();
+    std::swap(*data_.rbegin(), *(data_.begin() + offset));
+    data_.pop_back();
+    return iterator{data_.begin() + offset};
+  }
+
+  std::size_t erase(const K& key) {
+    for (auto it = data_.begin(); it != data_.end(); ++it) {
+      if (it->first == key) {
+        std::swap(*data_.rbegin(), *it);
+        data_.pop_back();
+        return 1;
+      }
+    }
+    return 0;
+  }
+
+  std::pair<iterator, bool> insert(const value_type& value) {
+    auto it = find(value.first);
+    if (it != end()) {
+      return {it, false};
+    }
+    data_.push_back(value);
+    return {iterator{data_.begin() + data_.size() - 1}, true};
+  }
+
+  template<typename M>
+  std::pair<iterator, bool> insert_or_assign(const K& key, M&& value) {
+    auto it = find(key);
+    if (it != end()) {
+      it->second = std::forward<M>(value);
+      return {it, false};
+    }
+    data_.emplace_back(key, std::forward<M>(value));
+    return {iterator{data_.begin() + data_.size() - 1}, true};
+  }
+
+  iterator find(const K& key) {
+    for (auto it = data_.begin(); it != data_.end(); ++it) {
+      if (it->first == key) return iterator{it};
+    }
+    return end();
+  }
+
+  const_iterator find(const K& key) const {
+    for (auto it = data_.begin(); it != data_.end(); ++it) {
+      if (it->first == key) return const_iterator{it};
+    }
+    return end();
+  }
+
+  iterator begin() noexcept {
+    return iterator{data_.begin()};
+  }
+
+  iterator end() noexcept {
+    return iterator{data_.end()};
+  }
+
+  const_iterator begin() const noexcept {
+    return const_iterator{data_.begin()};
+  }
+
+  const_iterator end() const noexcept {
+    return const_iterator{data_.end()};
+  }
+
+  const_iterator cbegin() const noexcept {
+    return const_iterator{data_.begin()};
+  }
+
+  const_iterator cend() const noexcept {
+    return const_iterator{data_.end()};
+  }
+
+  bool operator==(const FlatMap& other) const {
+    if (size() != other.size()) {
+      return false;
+    }
+    // no linearity can be guaranteed unless we
+    // sort the underlying storage
```

Comment:
Boost flat_map provides this guarantee by keeping the underlying vector sorted. Just FYI, not a change request.
https://www.boost.org/doc/libs/1_74_0/doc/html/boost/container/flat_map.html